update-client-hub/modules/pal-psa/source/arm_uc_pal_psa_implementation.c

```diff
@@ -0,0 +1,548 @@
+        UC_PAAL_TRACE("img: %" PRIX32, header_candidate->ih_img_size);
+        UC_PAAL_TRACE("prot: %" PRIX16, header_candidate->ih_protect_tlv_size);
+
+        /* Write details to KCM. The active header hash is used to identify
+         * which key-value pair to replace. */
+        result = arm_uc_pal_psa_set_kcm_details(&header_hash_active,
+                                                           header_hash_candidate,
+                                                           details_candidate);
+        if (result.error != ERR_NONE) {
+            return result;
+        }
+
+        psa_status_t status = psa_fwu_install(firmware_update_ctx_nonsecure.psa_fwu.image_id,
+                                              &firmware_update_ctx_nonsecure.psa_fwu.dependency_uuid,
+                                              &firmware_update_ctx_nonsecure.psa_fwu.dependency_version);
+        if (status != PSA_SUCCESS && status != PSA_SUCCESS_REBOOT) {
+            UC_PAAL_ERR_MSG("psa_fwu_install() failed: %d\r\n", status);
+            result.code = FIRM_ERR_ACTIVATE;
+            return result;
+        }
+
+        /* signal done */
+        arm_uc_pal_psa_signal_internal(ARM_UC_PAAL_EVENT_ACTIVATE_DONE);
+    
+        result.code = ERR_NONE;
+    }
+
+    return result;
+}
+
+/**
+ * @brief Get firmware details for the firmware image in the slot passed.
+ * @details This call populates the passed details struct with information
+ *          about the firmware image in the slot passed. Only the fields
+ *          marked as supported in the capabilities bitmap will have valid
+ *          values.
+ *
+ * @param details Pointer to firmware details struct to be populated.
+ * @return Returns ERR_NONE on accept, and signals the event handler with
+ *         either DONE or ERROR when complete.
+ *         Returns ERR_INVALID_PARAMETER on reject, and no signal is sent.
+ */
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_GetFirmwareDetails(
+    uint32_t slot_id,
+    arm_uc_firmware_details_t* details)
+{
+    UC_PAAL_TRACE("ARM_UC_PAL_PSA_Mcuboot_GetFirmwareDetails");
+
+    (void) slot_id;
+    (void) details;
+
+    /* This function is not used for MCUBOOT. */
+
+    arm_uc_error_t result = { .code = ERR_NOT_READY };
+
+    return result;
+}
+
+/*****************************************************************************/
+
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_GetActiveDetails(arm_uc_firmware_details_t* details)
+{
+    UC_PAAL_TRACE("ARM_UC_PAL_PSA_Mcuboot_GetActiveDetails");
+
+    arm_uc_error_t result = { .code = ERR_INVALID_PARAMETER };
+
+    if (details) {
+
+        /* Clear for clean */
+        memset(details, 0x00, sizeof(arm_uc_firmware_details_t));
+
+        /* parse MCUBOOT header and get hash from TLV struct and total size */
+        arm_uc_hash_t header_hash = { 0 };
+        uint32_t total_size = 0;
+
+        result = 
+            arm_uc_pal_psa_get_hash_from_header(arm_uc_flashiap_active_read,
```

Comment:
I was expecting the hash to be read using the PSA call `psa_fwu_query` and not directly from memory: https://github.com/ARMmbed/mbed-os/blob/master/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/include/psa/update.h#L178-L193

Reply:
Originally, I also expect `psa_fwu_query()` can do the hash calculation, but situation isn't so smooth: TF-M's `psa_fwu_query()` doesn't return hash of active firmware for security concern
https://github.com/OpenNuvoton/trusted-firmware-m/blob/149271c63a448b0d690b41ae62dfbf3f36abcf72/secure_fw/partitions/firmware_update/bootloader/mcuboot/tfm_mcuboot_fwu.c#L529-L533

Reply:
Sounds good, I'll take another look and get back to you!